src/goto-symex/goto_symex_state.cpp

```diff
@@ -438,19 +441,20 @@ bool goto_symex_statet::l2_thread_read_encoding(
 
     const bool record_events_bak=record_events;
     record_events=false;
-    assignment(ssa_l1, tmp, ns, true, true);
+    ssa_exprt ssa_l2 = assignment(std::move(ssa_l1), tmp, ns, true, true).get();
     record_events=record_events_bak;
 
     symex_target->assignment(
       guard_as_expr,
-      ssa_l1,
-      ssa_l1,
-      ssa_l1.get_original_expr(),
+      ssa_l2,
+      ssa_l2,
+      ssa_l2.get_original_expr(),
       tmp,
       source,
       symex_targett::assignment_typet::PHI);
 
-    expr = set_indices<L2>(std::move(ssa_l1), ns).get();
+    // TODO: are we setting l2 indices of something that is already l2?
+    expr = set_indices<L2>(std::move(ssa_l2), ns).get();
```

Comment:
Could you try adding an assertion that this effectively returns `ssa_l2` - I agree that this seems unnecessary, but maybe we're missing something.

Reply:
I've now added an invariant before the call to check that ssa_l2 was correctly renamed.